[BUG] ES256 signer is not according to spec

**Describe the bug**
The ES256 implementation in `APNSwiftSigner` stores the signature in the `DEM` format. According to the [spec](https://tools.ietf.org/html/rfc7515#appendix-A.3) the signature should be the concatenation of the `R` and `S` values as big-endian integers.

It's funny, because Apple's server actually accepts these JWTs. When you try to check a JWT generated from this library with [jwt.io](https://jwt.io) you will however get an "Invalid Signature" error.

If you change the code to produce the correct format, Apple's server is still happy. So I'm not sure what the best course of action is? Probably would be better to use the correct encoding, wdyt?

The correct encoding would be something like this:

```swift
let sig = try digest.withUnsafeBytes { ptr -> UnsafeMutablePointer<ECDSA_SIG> in
    guard let sig = ECDSA_do_sign(ptr.baseAddress?.assumingMemoryBound(to: UInt8.self), Int32(ptr.count), pkey) else {
        throw SigningError("Error while signing.")
    }
    return sig
}
defer { ECDSA_SIG_free(sig) }

let r = sig.pointee.r
let s = sig.pointee.s

var rb = [UInt8](repeating: 0, count: Int(BN_num_bits(r)+7)/8)
var sb = [UInt8](repeating: 0, count: Int(BN_num_bits(s)+7)/8)
let lenr = Int(BN_bn2bin(r, &rb))
let lens = Int(BN_bn2bin(s, &sb))

return Array(rb[0..<lenr] + sb[0..<lens])
```

### Example

Generate a new key pair:  
`openssl ecparam -genkey -name prime256v1 -noout -out ec256-key-pair.pem`

**Private Key**
```
-----BEGIN EC PRIVATE KEY-----
MHcCAQEEIEY5/amzr1QgHrLNZ8eHu926YERGWqB6QaDpNFcGxjsToAoGCCqGSM49
AwEHoUQDQgAEdbP7WQ/U4e5/CAqoBxatQb/5CEgJ070yMNGmWg5O6v2Q4M0l4CXK
cc94a66VttRZgVg6jE/ju+2mdHP7JWLmcQ==
-----END EC PRIVATE KEY-----
```

**Public Key**
```
-----BEGIN PUBLIC KEY-----
MFkwEwYHKoZIzj0CAQYIKoZIzj0DAQcDQgAEdbP7WQ/U4e5/CAqoBxatQb/5CEgJ
070yMNGmWg5O6v2Q4M0l4CXKcc94a66VttRZgVg6jE/ju+2mdHP7JWLmcQ==
-----END PUBLIC KEY-----
```

Using this library produces for example this JWT:

```
eyJraWQiOiJEOUFZSVhJM0ZKIiwiYWxnIjoiRVMyNTYifQ.eyJpc3MiOiJQSVo4SEpOVkJCIiwiaWF0IjoxNTY1MjU5NjQxfQ.MEUCIBWC-39SmgpRt_u6KJoroxahLprEo0cSLCoTFQOuXF6SAiEAxkpZob5j4RioIuW323lUAAAqbXE7E1A5JbHshjsMMiQ
```

If you copy this and try to verify it in jwt.io, it will fail.

With the encoding shown above you will get the following JWT. This JWT verifies correctly on jwt.io

```
eyJraWQiOiJEOUFZSVhJM0ZKIiwiYWxnIjoiRVMyNTYifQ.eyJpc3MiOiJQSVo4SEpOVkJCIiwiaWF0IjoxNTY1MjU5NjQxfQ.Mt2H-xWK4Z_at-thGusdit-eOYa9EksMu_LvVBa1irvQ7qfuZ1KcqH61d48eOI2ZaK_cgGNLAijwg3f2WF_mJA
```

```
/code/Sources/APNSwift/APNSwiftSigner.swift:48:84: error: use of undeclared type 'ECDSA_SIG'
        let sig = try digest.withUnsafeReadableBytes { ptr -> UnsafeMutablePointer<ECDSA_SIG> in
                                                                                   ^~~~~~~~~
/code/Sources/APNSwift/APNSwiftSigner.swift:48:54: error: cannot convert value of type '(_) throws -> <<error type>>' to expected argument type '(UnsafeRawBufferPointer) throws -> _'
        let sig = try digest.withUnsafeReadableBytes { ptr -> UnsafeMutablePointer<ECDSA_SIG> in
                                                     ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```
Now were just back to having this error.

--- a/Dockerfile
+++ b/Dockerfile
@@ -1,7 +1,9 @@
 FROM swift:5.0
 
 WORKDIR /code
-RUN apt-get install openssl libssl-dev
+RUN export DEBIAN_FRONTEND=noninteractive DEBCONF_NONINTERACTIVE_SEEN=true && apt-get -q update && \
+    apt-get -q install -y \
+    openssl libssl-dev
 COPY Package.swift /code/.
 RUN swift package resolve
 COPY ./Sources /code/Sources

--- a/Sources/APNSwift/APNSwiftSigner.swift
+++ b/Sources/APNSwift/APNSwiftSigner.swift
@@ -45,26 +45,28 @@ public struct APNSwiftSigner {
         }
         defer { EC_KEY_free(opaquePointer) }
         
-        let sig = try digest.withUnsafeReadableBytes { ptr -> UnsafeMutablePointer<ECDSA_SIG> in
+        let sig = try digest.withUnsafeReadableBytes { ptr -> OpaquePointer in
             guard let sig = ECDSA_do_sign(ptr.baseAddress?.assumingMemoryBound(to: UInt8.self), Int32(digest.readableBytes), opaquePointer) else {
                 throw APNSwiftError.SigningError.invalidSignatureData
             }
-            return sig
+            return .init(sig)
         }
-        defer { ECDSA_SIG_free(sig) }
+        defer { ECDSA_SIG_free(.init(sig)) }
 
-        let r = sig.pointee.r
-        let s = sig.pointee.s
-
+        var r : UnsafePointer<BIGNUM>? = nil
+        var s : UnsafePointer<BIGNUM>? = nil
+        
+        // as this method is `get0` there is no requirement to free those pointers: ECDSA_SIG will free them for us.
+        CAPNSOpenSSL_ECDSA_SIG_get0(.init(sig), &r, &s)
+        
         var rb = [UInt8](repeating: 0, count: Int(BN_num_bits(r)+7)/8)
         var sb = [UInt8](repeating: 0, count: Int(BN_num_bits(s)+7)/8)
         let lenr = Int(BN_bn2bin(r, &rb))
         let lens = Int(BN_bn2bin(s, &sb))
 
-        let finalSig = Array(rb[0..<lenr] + sb[0..<lens])
-        
-        var derBytes = ByteBufferAllocator().buffer(capacity: Int(finalSig.count))
-        derBytes.writeBytes(UnsafeBufferPointer<CUnsignedChar>(start: finalSig, count: Int(finalSig.count)))
-        return derBytes
+        var signatureBytes = ByteBufferAllocator().buffer(capacity: lenr + lens)
+        signatureBytes.writeBytes(rb[0..<lenr])
+        signatureBytes.writeBytes(sb[0..<lens])
+        return signatureBytes
     }
 }

--- a/Sources/APNSwift/APNSwiftSigner.swift
+++ b/Sources/APNSwift/APNSwiftSigner.swift
@@ -46,7 +46,7 @@ public struct APNSwiftSigner {
         defer { EC_KEY_free(opaquePointer) }
         
         let sig = try digest.withUnsafeReadableBytes { ptr -> OpaquePointer in
-            guard let sig = ECDSA_do_sign(ptr.baseAddress?.assumingMemoryBound(to: UInt8.self), Int32(digest.readableBytes), opaquePointer) else {
+            guard let sig = ECDSA_do_sign(ptr.baseAddress?.assumingMemoryBound(to: UInt8.self), Int32(ptr.count), opaquePointer) else {
                 throw APNSwiftError.SigningError.invalidSignatureData
             }
             return .init(sig)

--- a/Sources/APNSwift/APNSwiftSigner.swift
+++ b/Sources/APNSwift/APNSwiftSigner.swift
@@ -45,19 +45,26 @@ public struct APNSwiftSigner {
         }
         defer { EC_KEY_free(opaquePointer) }
         
-        let sig = digest.withUnsafeReadableBytes { ptr in
-            ECDSA_do_sign(ptr.baseAddress?.assumingMemoryBound(to: UInt8.self), Int32(digest.readableBytes), opaquePointer)
+        let sig = try digest.withUnsafeReadableBytes { ptr -> UnsafeMutablePointer<ECDSA_SIG> in
+            guard let sig = ECDSA_do_sign(ptr.baseAddress?.assumingMemoryBound(to: UInt8.self), Int32(digest.readableBytes), opaquePointer) else {
+                throw APNSwiftError.SigningError.invalidSignatureData
+            }
+            return sig
         }
         defer { ECDSA_SIG_free(sig) }
 
-        var derEncodedSignature: UnsafeMutablePointer<CUnsignedChar>?
-        let derLength = i2d_ECDSA_SIG(sig, &derEncodedSignature)
-        guard let derCopy = derEncodedSignature, derLength > 0 else {
-            throw APNSwiftError.SigningError.invalidASN1
-        }
+        let r = sig.pointee.r
+        let s = sig.pointee.s
 
-        var derBytes = ByteBufferAllocator().buffer(capacity: Int(derLength))
-        derBytes.writeBytes(UnsafeBufferPointer<CUnsignedChar>(start: derCopy, count: Int(derLength)))
+        var rb = [UInt8](repeating: 0, count: Int(BN_num_bits(r)+7)/8)
+        var sb = [UInt8](repeating: 0, count: Int(BN_num_bits(s)+7)/8)
+        let lenr = Int(BN_bn2bin(r, &rb))
+        let lens = Int(BN_bn2bin(s, &sb))
+
+        let finalSig = Array(rb[0..<lenr] + sb[0..<lens])
+        
+        var derBytes = ByteBufferAllocator().buffer(capacity: Int(finalSig.count))
+        derBytes.writeBytes(UnsafeBufferPointer<CUnsignedChar>(start: finalSig, count: Int(finalSig.count)))
         return derBytes
     }
 }

--- a/Sources/CAPNSOpenSSL/shim.h
+++ b/Sources/CAPNSOpenSSL/shim.h
@@ -9,6 +9,18 @@
 #include <openssl/crypto.h>
 
 
+/// ECDSA_SIG_get0() returns internal pointers the r and s values contained in
+/// sig and stores them in *pr and *ps, respectively. The pointer pr or ps can
+/// be NULL, in which case the corresponding value is not returned.
+static inline void CAPNSOpenSSL_ECDSA_SIG_get0(const ECDSA_SIG *sig, const BIGNUM **pr, const BIGNUM **ps) {
+#if (OPENSSL_VERSION_NUMBER < 0x10100000L) || defined(LIBRESSL_VERSION_NUMBER)
+        if (pr != NULL)
+            *pr = sig->r;
+        if (ps != NULL)
+            *ps = sig->s;
+#else
+        ECDSA_SIG_get0(sig, pr, ps);
 #endif
+}
 
-
+#endif

--- a/Tests/APNSwiftJWTTests/JWTTests.swift
+++ b/Tests/APNSwiftJWTTests/JWTTests.swift
@@ -14,6 +14,7 @@
 
 import Foundation
 import XCTest
+import NIO
 @testable import APNSwift
 
 final class JWTTests: XCTestCase {
@@ -53,8 +54,32 @@ final class JWTTests: XCTestCase {
             XCTFail("Payload can't be decoded")
         }
     }
+    
+    func testJWTSigning() throws {
+        let teamID = "8RX5AF8F6Z"
+        let keyID = "9N8238KQ6Z"
+        let date = Date()
+        let jwt = APNSwiftJWT(keyID: keyID, teamID: teamID, issueDate: date, expireDuration: 10.0)
+        
+        let privateKey = """
+-----BEGIN EC PRIVATE KEY-----
+MHcCAQEEIEY5/amzr1QgHrLNZ8eHu926YERGWqB6QaDpNFcGxjsToAoGCCqGSM49
+AwEHoUQDQgAEdbP7WQ/U4e5/CAqoBxatQb/5CEgJ070yMNGmWg5O6v2Q4M0l4CXK
+cc94a66VttRZgVg6jE/ju+2mdHP7JWLmcQ==
+-----END EC PRIVATE KEY-----
+"""
+        
+        var buffer = ByteBufferAllocator().buffer(capacity: privateKey.count)
+        buffer.writeString(privateKey)
+        
+        let signer = try APNSwiftSigner(buffer: buffer)
+        let sig = try signer.sign(digest: try jwt.getDigest().fixedDigest)
+        XCTAssertEqual(sig.readableBytes, 64) // len(r) + len(s) == 64 byte
+    }
+    
     static var allTests = [
         ("testJWTEncodingAndSign", testJWTEncoding),
+        ("testJWTSigning", testJWTSigning),
     ]
 }
 

